gui/dig.lua

```diff
@@ -0,0 +1,658 @@
+                                    (self.shape.options[key].value <= self.shape.options[key].max)
+                            end,
+                            disabled = false,
+                            show_tooltip = true,
+                            on_activate = function()
+                                self.shape.options[key].value =
+                                self.shape.options[key].value + 1
+                                self.dirty = true
+                            end,
+                        } }
+                end
+            end
+        end
+    end
+end
+
+function Dig:save_cursor_pos()
+    self.saved_cursor = copyall(df.global.cursor)
+end
+
+function Dig:on_mark(pos)
+    self.mark = pos
+    self:updateLayout()
+end
+
+function Dig:get_bounds()
+    local cur = self.saved_cursor or dfhack.gui.getMousePos()
+    if not cur then return end
+    local mark = self.mark or cur
+
+    return {
+        x1 = math.min(cur.x, mark.x),
+        x2 = math.max(cur.x, mark.x),
+        y1 = math.min(cur.y, mark.y),
+        y2 = math.max(cur.y, mark.y),
+        z1 = math.min(cur.z, mark.z),
+        z2 = math.max(cur.z, mark.z),
+    }
+end
+
+function Dig:onRenderFrame(dc, rect)
+    Dig.super.onRenderFrame(self, dc, rect)
+    if self.shape == nil then
+        self.shape =
+        shapes.all_shapes[self.subviews.shape_name:getOptionValue()]
+    end
+
+    local bounds = self:get_bounds()
+    if bounds and self.mark then
+
+        local function get_overlay_pen(pos)
+            -- Check if we need to update the shape dimensions. Either there isn't a shape, we've marked it dirty, or the bounds have changed
+            if self.dirty or (pos.x >= bounds.x1 and pos.x <= bounds.x2 and pos.y >= bounds.y1 and pos.y <= bounds.y2) then
+                if not self.shape or self.dirty or
+                    (bounds.x2 - bounds.x1 ~= self.shape.width or bounds.y2 - bounds.y1 ~= self.shape.height) then
+                    self.shape:update(
+                        bounds.x2 - bounds.x1,
+                        bounds.y2 - bounds.y1
+                    )
+                    self.dirty = false
+                end
+
+                local mouse_pos = dfhack.gui.getMousePos()
+                if mouse_pos ~= nil then
+                    mouse_pos.x = mouse_pos.x - bounds.x1
+                    mouse_pos.y = mouse_pos.y - bounds.y1
+                end
+
+                -- Get the pen from the base Shape class based on if the point is in the shape or not
+                -- Send mouse position for stuff like corner anchor mouse over, etc...
+                return self.shape:get_pen(
+                    pos.x - bounds.x1,
+                    pos.y - bounds.y1,
+                    mouse_pos
+                )
+            else
+                return nil
+            end
+        end
+
+        guidm.renderMapOverlay(get_overlay_pen, bounds)
```

Comment:
another option is to factor the preview display logic out from gui/quickfort and use that, but this isn't complicated enough yet to go that route, I think. Maybe once you implement 3d shapes then there will be a better reason to share logic.